IndexedDB/idbcursor_advance_index2.htm

```diff
@@ -34,7 +34,7 @@
         cursor_rq.onsuccess = t.step_func(function(e) {
             var cursor = e.target.result;
 
-            assert_true(cursor != null, "cursor exist");
+            assert_not_equals(cursor, null, "cursor exist");
```

Comment:
This is replacing `!=`, not `!==` like the commit message states.

`x != null` is a [common pattern](https://twitter.com/getify/status/1274454560610889730) to check valid inputs including `""`, `0` and `false`, but not `null` or `undefined`.

This patch effectively replaces `!=` with `!==`, flipping treatment of all `undefined` values tested.

Reply:
In the case of an IDBRequest's success event for a cursor request (opening a cursor or iterating the cursor), the `result` property is `null` when no more results are present. So "not null" is semantically correct here from the perspective of the Indexed DB spec. 

Agreed, though, that it technically reduces coverage, if an implementation provided undefined here. Of course, the test as written would pass if practically any value was provided e.g. "something" is both not null and will result in a TypeError in the subsequent assertion.

A better assertion would be that `result` is an instance of `IDBCursorWithValue`

Reply:
FWIW: there's a reason why the `==`/`!=` changes were in separate commits; the commit messages should be accurate. Note that, as noted above, `===`/`!==` is also a behavioural change (our asserts are comparable to `Object.is`).

We really don't want to be allowing everything `x != null` in the test suite: I'm unaware of _any_ spec that allows implementations to use `null` _or_ `undefined` as a given return value. We really do want to be as specific as possible when checking return values.

Are there risks here? Yes. That we don't have any regressions in Chrome/Firefox/Safari when running any of these tests is a good sign that tests do mean the specific value, rather than relying on the coercion between types. While I've not looked in depth at the changes, I certainly can't see any case which appears to be relying on type coercion in the comparison.

Reply:
@gsnedders I agree accuracy in tests is tantamount, which is why I question a semantics-changing regex sweep.

Here's what I recommend. Replace
```js
assert_true(x != null, "writer's intention");
```
with
```js
assert_not_equals(x, null, "writer's intention");
assert_not_equals(x, undefined, "writer's intention");
```
Then leave it to individual domain experts to remove redundancy. This makes it behavior neutral and explicit.

And thanks for doing this! Being able to see failing values in asserts saves so much time. 👍

Reply:
FWIW my opinion is this is overkill. If we have made the change and don't see any test result changes across multiple browsers, and don't know of specific cases which we believe are incorrect, the probability of introducing a test bug this way seems acceptably low and in particular is likely no higher than the probability of a test being buggy for many other reasons (e.g. only checked in a single browser to begin with).

I also agree with the point made by @inexorabletash that using `!=[=]` in tests is pretty sketchy from a correctness point of view and we'd almost always be better off making a positive assert on the actual expected value or type. If we're going to spend more effort here let's make that change instead.

Reply:
@jgraham for `assert_true(x != null)` → `assert_not_equals(x, null)` we wouldn't see test result failures becaise it's more permissive not less.

Reply:
But we could see new passes. I'm not claiming there's zero risk here, but I don't think the risk is enough to block the PR.

Reply:
@jgraham I wouldn't expect new passes, since `expected: FAIL` tests tend to represent missing functionality, so having one accidentally pass from this change could only happen if the test was broken I think. Those tests are also a separate and much smaller sub-population I think.

Instead of looking at this in terms of fail/pass, why not preserve the current accuracy? We know all tests pass under the older stricter requirements, so why not preserve those requirements by adding the extra asserts?

Reply:
`grep -r "!= null" * | grep assert_true` shows [108 hits](https://gist.github.com/jan-ivar/d2f1cd2b94f6d71f6323e34d57f0206c).

Reply:
For comparison, `rg 'assert_not_equals\(.+,\s*(null|undefined)\s*[,)]' --no-heading | wc -l` shows 647 hits on master and 791 on this branch, so in many ways this ship has already sailed. I don't totally oppose reverting the changes in this branch, and there's now a couple of fixup commits that get rid of them.